src/vga/vga_color.rs

```diff
@@ -1,95 +1,189 @@
+}
+
+impl<T> Sub for VGAColor<T>
+where
+    T: core::ops::Sub<Output = T>,
+{
+    type Output = VGAColor<T>;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        VGAColor {
+            red: self.red - rhs.red,
+            green: self.green - rhs.green,
+            blue: self.blue - rhs.blue,
+            alpha: self.alpha - rhs.alpha,
         }
     }
-    /// Returns the byte value for the grayscale version of the color, using the human light perception formula
-    pub fn to_grayscale(&self) -> u8 {
-        let red = self.red as u32;
-        let green = self.green as u32;
-        let blue = self.blue as u32;
-        ((red * 299 + green * 587 + blue * 114) / 1000) as u8
+}
+
+impl<T> Mul for VGAColor<T>
+where
+    T: core::ops::Mul<Output = T>,
+{
+    type Output = VGAColor<T>;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        VGAColor {
+            red: self.red * rhs.red,
+            green: self.green * rhs.green,
+            blue: self.blue * rhs.blue,
+            alpha: self.alpha * rhs.alpha,
+        }
     }
-    /// Interpolates between two colors, where t=0 -> First color, t=255 -> Second color
-    pub fn interpolate(a: &VGAColor, b: &VGAColor, t: u8) -> VGAColor {
-        let t1 = (255 - t) as u16;
-        let t2 = t as u16;
+}
+
+impl<T> Mul<T> for VGAColor<T>
+where
+    T: core::ops::Mul<Output = T> + Copy,
+{
+    type Output = VGAColor<T>;
+
+    fn mul(self, rhs: T) -> Self::Output {
         VGAColor {
-            red: ((a.red as u16 * t1 + b.red as u16 * t2) / 255) as u8,
-            green: ((a.green as u16 * t1 + b.green as u16 * t2) / 255) as u8,
-            blue: ((a.blue as u16 * t1 + b.blue as u16 * t2) / 255) as u8,
-            alpha: ((a.alpha as u16 * t1 + b.alpha as u16 * t2) / 255) as u8,
+            red: self.red * rhs,
+            green: self.green * rhs,
+            blue: self.blue * rhs,
+            alpha: self.alpha * rhs,
+        }
+    }
+}
+
+impl<T> Div for VGAColor<T>
+where
+    T: core::ops::Div<Output = T>,
+{
+    type Output = VGAColor<T>;
+
+    fn div(self, rhs: Self) -> Self::Output {
+        VGAColor {
+            red: self.red / rhs.red,
+            green: self.green / rhs.green,
+            blue: self.blue / rhs.blue,
+            alpha: self.alpha / rhs.alpha,
+        }
+    }
+}
+
+impl<T> Div<T> for VGAColor<T>
+where
+    T: core::ops::Div<Output = T> + Copy,
+{
+    type Output = VGAColor<T>;
+
+    fn div(self, rhs: T) -> Self::Output {
+        VGAColor {
+            red: self.red / rhs,
+            green: self.green / rhs,
+            blue: self.blue / rhs,
+            alpha: self.alpha / rhs,
+        }
+    }
+}
+
+impl<T> VGAColor<T>
+where
+    T: Into<u32> + Copy,
+{
+    /// Returns the value for the grayscale version of the color, using the human light perception formula
+    pub fn to_grayscale(self) -> u32 {
+        (self.red.into() * 299 + self.green.into() * 587 + self.blue.into() * 114) / 1000
+    }
+}
+
+impl VGAColor<u8> {
```

Comment:
why is this only implemented for `u8`

Reply:
Because the opacity is passed as u8, not as f32, i.e. it's 0-255, there's no point in implementing this for other sub-types, as it would be a pain really (how do you tell which of them to cast to u16 and which to leave be as they have bigger size?)